File: src/personaljarvis/base/owner_attestation.py
"""Der Beleg, dass der Eigentümer **diese** Mutation freigegeben hat.

**Warum es das gibt.** Bis hierher entstand eine Eigentümerentscheidung aus
einem HTTP-Aufruf mit frei gesetztem `decision_actor`. Gemessen am 2026-08-17:
25 von 25 Mutationen liefen so nacheinander durch, ohne dass ein Mensch etwas
bestätigte. Der Name im Aufruf war ein Metadatum und wurde als
Authentizitätsbeweis gelesen — das ist der Fehler, den diese Datei behebt.

**Was ein Beleg ist.** Eine Datei, die der **App-Prozess** nach einer
Eigentümerhandlung schreibt und die der Kern nur liest. Sie bindet sich an
genau eine Mutation und an genau deren Nutzlast:

* `mutation_id` — der Beleg gilt für diesen Vorgang und keinen anderen.
* `payload_digest` — ändert sich die Nutzlast nach der Handlung, passt der
  Beleg nicht mehr.
* `preview_digest` — und ändert sich die **Darstellung**, passt er ebenso
  wenig. Die Nutzlast allein genügt nicht: Derselbe Datensatz lässt sich
  verschieden zeigen, und freigegeben wurde, was auf der Fläche stand. Die
  Vorschau deckt seit 2026-08-13 auch Name und Art des Zielablageorts, also
  bindet der Beleg genau die sechs Angaben, die der Eigentümer gesehen hat.
* `attested_at` — er ist frisch oder er ist keiner. Ein alter Beleg darf keine
  spätere Mutation tragen.
* Einmalig — `consume()` entfernt ihn. Kein zweiter Lauf aus derselben
  Handlung.

**Was das ist und was nicht.** Dieselbe Semantik wie überall in diesem
Projekt: `requires_interactive_owner_authentication`, nicht
`technically_impossible`. Der Beleg liegt im Benutzerordner, und jeder Prozess
unter derselben Benutzerkennung kann ihn schreiben. Gesichert ist, dass es im
Kern, im Backend und in der HTTP-Schicht **keinen Codepfad gibt**, der einen
Beleg erzeugt — ein Statiktest hält das fest. Der Agent spricht über HTTP mit
dem Kern; er erreicht damit keinen Erzeugungsweg mehr.

**Gemeinsam, nicht je Fachbereich.** Die Fähigkeit steht im Beleg und wird
verglichen. Kalender wird dasselbe Primitiv verwenden; bis dahin verhindert
sein Produktschloss ohnehin jede Mutation.
"""

from __future__ import annotations

import json
import os
import stat
from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from pathlib import Path

__all__ = [
    "ATTESTATION_CONTRACT",
    "ATTESTATION_DIRNAME",
    "MAX_ATTESTATION_AGE_SECONDS",
    "OwnerAttestation",
    "default_attestation_dir",
    "read_attestation",
    "consume_attestation",
]

#: Vertragskennung. Ändert sich der Beleg, wird jeder alte dadurch wirkungslos.
ATTESTATION_CONTRACT = "owner-approval-v1"

#: Verzeichnis neben der Personal-Datenbank — dort, wo schon die
#: Schreibfreigabe liegt und wo ohnehin nur der Eigentümer hinkommt.
ATTESTATION_DIRNAME = "owner-approvals"

#: Wie alt ein Beleg höchstens sein darf. Bewusst kurz und bewusst dieselbe
#: Grössenordnung wie die Freigabefrist: Der Beleg gehört zur Handlung, nicht
#: zum Tag.
MAX_ATTESTATION_AGE_SECONDS = 15 * 60

#: Wie weit ein Beleg in der Zukunft liegen darf. Uhren gehen auseinander;
#: eine halbe Minute ist Nachsicht, fünf Minuten wären ein Schlupfloch.
_MAX_SKEW_SECONDS = 30


@dataclass(frozen=True, slots=True)
class OwnerAttestation:
    """Ein geprüfter, noch nicht verbrauchter Beleg."""

    capability: str
    mutation_id: str
    payload_digest: str
    preview_digest: str
    attested_at: str
    #: Wie der Eigentümer gehandelt hat. Protokoll, nicht Nachweis — der
    #: Nachweis ist, dass der Beleg überhaupt aus dem App-Prozess kommt.
    method: str


def default_attestation_dir(database_path: Path | str | None = None) -> Path:
    if database_path is not None:
        return Path(database_path).parent / ATTESTATION_DIRNAME
    from personaljarvis.base.db.factory import default_database_path

    return default_database_path().parent / ATTESTATION_DIRNAME


def _ist_privat(pfad: Path, *, verzeichnis: bool) -> bool:
    try:
        st = pfad.lstat()
    except OSError:
        return False
    if verzeichnis and not stat.S_ISDIR(st.st_mode):
        return False
    if not verzeichnis and (stat.S_ISLNK(st.st_mode) or not stat.S_ISREG(st.st_mode)):
        return False
    erwartet = 0o700 if verzeichnis else 0o600
    return st.st_uid == os.geteuid() and stat.S_IMODE(st.st_mode) == erwartet


def _pfad(verzeichnis: Path, mutation_id: str) -> Path | None:
    """Der Ort des Belegs — oder `None`, wenn die Kennung keiner ist.

    Die Kennung landet in einem Dateinamen. Ein Punkt oder ein Schrägstrich
    darin wäre ein Ausbruch aus dem Verzeichnis, deshalb wird sie hier
    ausdrücklich eingeschränkt statt bereinigt.
    """
    if not mutation_id or not all(z.isalnum() or z == "-" for z in mutation_id):
        return None
    return verzeichnis / f"{mutation_id}.json"


def read_attestation(*, capability: str, mutation_id: str, payload_digest: str,
                     preview_digest: str,
                     verzeichnis: Path | str | None = None,
                     jetzt: datetime | None = None) -> OwnerAttestation | None:
    """Liest den Beleg — oder gibt `None` zurück. Nie eine Ausnahme.

    Fail-closed und still: Ein fehlender Beleg ist der Normalfall (die meisten
    Vorgänge sind noch nicht freigegeben) und kein Fehler.
    """
    ort = Path(verzeichnis) if verzeichnis is not None else default_attestation_dir()
    if not _ist_privat(ort, verzeichnis=True):
        return None
    datei = _pfad(ort, mutation_id)
    if datei is None or not _ist_privat(datei, verzeichnis=False):
        return None
    try:
        roh = json.loads(datei.read_text(encoding="utf-8"))
    except (OSError, ValueError):
        return None
    if not isinstance(roh, dict):
        return None
    if roh.get("contract") != ATTESTATION_CONTRACT:
        return None
    if roh.get("capability") != capability:
        return None
    # Die Bindung. Ein Beleg für einen anderen Vorgang oder für eine andere
    # Nutzlast ist kein Beleg für diesen.
    if roh.get("mutation_id") != mutation_id:
        return None
    if roh.get("payload_digest") != payload_digest:
        return None
    # Und die Darstellung. Ohne sie deckte der Beleg dieselbe Nutzlast unter
    # einer anderen Anzeige — etwa mit einem anderen Zielablageort im Text.
    if roh.get("preview_digest") != preview_digest:
        return None

    gestempelt = roh.get("attested_at")
    if not isinstance(gestempelt, str):
        return None
    try:
        moment = datetime.fromisoformat(gestempelt.replace("Z", "+00:00"))
    except ValueError:
        return None
    if moment.tzinfo is None:
        moment = moment.replace(tzinfo=timezone.utc)

    now = jetzt or datetime.now(timezone.utc)
    if moment > now + timedelta(seconds=_MAX_SKEW_SECONDS):
        return None
    if moment < now - timedelta(seconds=MAX_ATTESTATION_AGE_SECONDS):
        return None

    verfahren = roh.get("method")
    if not isinstance(verfahren, str) or not verfahren.strip():
        return None

    return OwnerAttestation(
        capability=capability, mutation_id=mutation_id,
        payload_digest=payload_digest, preview_digest=preview_digest,
        attested_at=gestempelt, method=verfahren.strip())


def consume_attestation(*, mutation_id: str,
                        verzeichnis: Path | str | None = None) -> bool:
    """Entfernt den Beleg. Eine Handlung, eine Mutation.

    Bleibt er liegen, trüge dieselbe Eigentümerhandlung einen zweiten Lauf —
    und genau das ist die Schleife, gegen die dieses Modul steht.
    """
    ort = Path(verzeichnis) if verzeichnis is not None else default_attestation_dir()
    datei = _pfad(ort, mutation_id)
    if datei is None:
        return False
    try:
        datei.unlink()
        return True
    except OSError:
        return False

File: frontend/src-tauri/src/owner_attestation.rs
//! Der Schreiber des Eigentümerbelegs — die andere Hälfte von
//! `personaljarvis/base/owner_attestation.py`.
//!
//! **Warum hier.** Der Kern liest den Beleg, erzeugt ihn aber nie. Läge der
//! Erzeugungsweg im Python-Backend, könnte der Agent ihn über HTTP erreichen —
//! genau die Lücke, die am 2026-08-17 gemessen wurde: 25 von 25 Mutationen
//! liefen mit frei gesetztem `decision_actor` durch. Der Weg lebt deshalb im
//! App-Prozess, aufrufbar nur aus der Fläche, die der Eigentümer bedient.
//!
//! **Was der Beleg bindet.** Vorgang, Nutzlast **und Darstellung**. Die
//! Darstellung gehört dazu, weil der Eigentümer eine Fläche freigibt und
//! keinen Digest: Ändert sich danach die Nutzlast oder die Anzeige, passt der
//! Beleg nicht mehr, und der Kern lehnt ab.
//!
//! **Die Grenze, ausgesprochen.** Gesichert ist: es gibt im Kern, im Backend
//! und in der HTTP-Schicht keinen Codepfad, der einen Beleg erzeugt — der
//! Agent erreicht damit keinen Erzeugungsweg. Nicht gesichert ist, dass nur
//! der Eigentümer einen Beleg schreiben kann: Die Datei liegt in seinem
//! Benutzerordner, und jeder Prozess unter derselben Benutzerkennung kann sie
//! anlegen. Dieselbe Lage wie bei der Freigabedatei und beim Guard —
//! `requires_interactive_owner_authentication`, nicht
//! `technically_impossible`. Ein stärkerer Herkunftsnachweis gegen
//! gleichberechtigte lokale Prozesse ist eine eigene Aufgabe.

use std::io::Write;
use std::os::unix::fs::{MetadataExt, OpenOptionsExt, PermissionsExt};
use std::path::{Path, PathBuf};

use serde::Serialize;

/// Vertragskennung — wortgleich zu `ATTESTATION_CONTRACT`.
pub const ATTESTATION_CONTRACT: &str = "owner-approval-v1";
/// Verzeichnisname — wortgleich zu `ATTESTATION_DIRNAME`.
pub const ATTESTATION_DIRNAME: &str = "owner-approvals";

/// Wie der Eigentümer gehandelt hat. Protokoll, nicht Nachweis: Der Nachweis
/// ist, dass der Beleg überhaupt aus diesem Prozess kommt.
const METHOD_APP_INTERACTION: &str = "app_interaction";

#[derive(Debug, Serialize)]
pub struct AttestationResult {
    /// Ob ein gültiger Beleg jetzt liegt.
    pub written: bool,
    /// Geschlossene Kennung im Fehlerfall: `attestation_directory_unavailable`,
    /// `invalid_binding`, `write_failed`.
    pub reason_code: Option<String>,
}

impl AttestationResult {
    fn fehler(code: &str) -> Self {
        Self { written: false, reason_code: Some(code.into()) }
    }
}

fn attestation_dir() -> PathBuf {
    let home = std::env::var("OPENJARVIS_HOME")
        .map(PathBuf::from)
        .unwrap_or_else(|_| {
            PathBuf::from(std::env::var("HOME").unwrap_or_default()).join(".openjarvis")
        });
    home.join("personal").join(ATTESTATION_DIRNAME)
}

/// Eine Kennung, die in einen Dateinamen darf.
///
/// Eingeschränkt statt bereinigt — dieselbe Regel wie im Kern. Ein Punkt oder
/// ein Schrägstrich wäre ein Ausbruch aus dem Verzeichnis.
fn kennung_ist_zulaessig(wert: &str) -> bool {
    !wert.is_empty()
        && wert.len() <= 64
        && wert.chars().all(|z| z.is_ascii_alphanumeric() || z == '-')
}

fn ist_hex64(wert: &str) -> bool {
    wert.len() == 64 && wert.chars().all(|z| z.is_ascii_hexdigit() && !z.is_ascii_uppercase())
}

/// Legt das Belegverzeichnis an — 0700 vom Anlegen an.
///
/// Der Kern verlangt genau diese Rechte und lehnt sonst ab. Ein Verzeichnis
/// erst anzulegen und dann zu verschärfen liesse einen Augenblick offen.
fn verzeichnis_bereit(ort: &Path) -> bool {
    if let Ok(meta) = std::fs::symlink_metadata(ort) {
        return meta.is_dir()
            && meta.permissions().mode() & 0o777 == 0o700
            && meta.uid() == unsafe { libc::geteuid() };
    }
    let Some(eltern) = ort.parent() else {
        return false;
    };
    if !eltern.is_dir() {
        return false;
    }
    if std::fs::create_dir(ort).is_err() {
        return false;
    }
    std::fs::set_permissions(ort, std::fs::Permissions::from_mode(0o700)).is_ok()
}

fn urkunde(mutation_id: &str, payload_digest: &str, preview_digest: &str,
           jetzt_iso: &str) -> String {
    format!(
        concat!(
            "{{\n",
            "  \"contract\": \"{contract}\",\n",
            "  \"capability\": \"contacts\",\n",
            "  \"mutation_id\": \"{mutation}\",\n",
            "  \"payload_digest\": \"{payload}\",\n",
            "  \"preview_digest\": \"{preview}\",\n",
            "  \"attested_at\": \"{jetzt}\",\n",
            "  \"method\": \"{method}\"\n",
            "}}\n"
        ),
        contract = ATTESTATION_CONTRACT,
        mutation = mutation_id,
        payload = payload_digest,
        preview = preview_digest,
        jetzt = jetzt_iso,
        method = METHOD_APP_INTERACTION,
    )
}

fn schreiben(ort: &Path, inhalt: &str) -> std::io::Result<()> {
    let temporaer = ort.with_extension("json.neu");
    {
        let mut datei = std::fs::OpenOptions::new()
            .write(true)
            .create(true)
            .truncate(true)
            .mode(0o600)
            .open(&temporaer)?;
        datei.write_all(inhalt.as_bytes())?;
        datei.sync_all()?;
    }
    std::fs::rename(&temporaer, ort)
}

/// Der prüfbare Kern, ohne Tauri und ohne Uhr.
pub fn attestieren_in(verzeichnis: &Path, mutation_id: &str, payload_digest: &str,
                      preview_digest: &str, jetzt_iso: &str) -> AttestationResult {
    // Die Bindung muss vollstaendig sein, bevor irgendetwas entsteht. Ein
    // Beleg ohne sie waere ein Blankoscheck.
    if !kennung_ist_zulaessig(mutation_id)
        || !ist_hex64(payload_digest)
        || !ist_hex64(preview_digest)
    {
        return AttestationResult::fehler("invalid_binding");
    }
    if !verzeichnis_bereit(verzeichnis) {
        return AttestationResult::fehler("attestation_directory_unavailable");
    }
    let ort = verzeichnis.join(format!("{mutation_id}.json"));
    if schreiben(&ort, &urkunde(mutation_id, payload_digest, preview_digest, jetzt_iso))
        .is_err()
    {
        return AttestationResult::fehler("write_failed");
    }
    AttestationResult { written: true, reason_code: None }
}

// ── Tauri-Kommando ──────────────────────────────────────────────────────────

/// Die Eigentümerhandlung: Der Aufruf **ist** die Zustimmung.
///
/// Er ist ausschliesslich aus der Freigabefläche des App-Prozesses erreichbar.
/// Eine zusätzliche Systemauthentifizierung verlangt er bewusst nicht — der
/// sicherheitsentscheidende Punkt ist die Herkunft der Handlung, nicht
/// möglichst häufiges Fingerauflegen. Der Übergang der **Dauerfreigabe** von
/// aus auf an verlangt sie weiterhin.
#[tauri::command]
pub fn personal_contacts_attest_owner_approval(
    mutation_id: String,
    payload_digest: String,
    preview_digest: String,
) -> AttestationResult {
    attestieren_in(
        &attestation_dir(),
        &mutation_id,
        &payload_digest,
        &preview_digest,
        &crate::contacts_standing_write::iso_jetzt_oeffentlich(),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    const MUTATION: &str = "11111111-1111-4111-8111-111111111111";
    const PAYLOAD: &str = "a1b2c3d4e5f60718293a4b5c6d7e8f90a1b2c3d4e5f60718293a4b5c6d7e8f90";
    const PREVIEW: &str = "0f1e2d3c4b5a69788796a5b4c3d2e1f00f1e2d3c4b5a69788796a5b4c3d2e1f0";
    const JETZT: &str = "2026-08-17T12:00:00Z";

    fn ordner(name: &str) -> PathBuf {
        let dir = std::env::temp_dir().join(format!("jc-beleg-{}-{}", name, std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        std::fs::create_dir_all(&dir).unwrap();
        dir.join("owner-approvals")
    }

    #[test]
    fn der_beleg_traegt_vorgang_nutzlast_und_darstellung() {
        let ort = ordner("vollstaendig");
        let ergebnis = attestieren_in(&ort, MUTATION, PAYLOAD, PREVIEW, JETZT);
        assert!(ergebnis.written, "{:?}", ergebnis.reason_code);

        let inhalt = std::fs::read_to_string(ort.join(format!("{MUTATION}.json"))).unwrap();
        for erwartet in [MUTATION, PAYLOAD, PREVIEW, ATTESTATION_CONTRACT, "contacts"] {
            assert!(inhalt.contains(erwartet), "fehlt: {erwartet}");
        }
    }

    #[test]
    fn der_beleg_liegt_von_anfang_an_mit_0600_in_einem_0700_ordner() {
        let ort = ordner("rechte");
        assert!(attestieren_in(&ort, MUTATION, PAYLOAD, PREVIEW, JETZT).written);
        let datei = std::fs::metadata(ort.join(format!("{MUTATION}.json")))
            .unwrap()
            .permissions()
            .mode()
            & 0o777;
        let verzeichnis = std::fs::metadata(&ort).unwrap().permissions().mode() & 0o777;
        assert_eq!(datei, 0o600, "Datei {datei:o}");
        assert_eq!(verzeichnis, 0o700, "Verzeichnis {verzeichnis:o}");
    }

    #[test]
    fn eine_unvollstaendige_bindung_erzeugt_keinen_beleg() {
        let ort = ordner("bindung");
        for (mutation, payload, preview) in [
            ("", PAYLOAD, PREVIEW),
            ("../ausbruch", PAYLOAD, PREVIEW),
            ("a/b", PAYLOAD, PREVIEW),
            (MUTATION, "zu-kurz", PREVIEW),
            (MUTATION, PAYLOAD, "zu-kurz"),
            (MUTATION, &PAYLOAD.to_uppercase(), PREVIEW),
        ] {
            let ergebnis = attestieren_in(&ort, mutation, payload, preview, JETZT);
            assert!(!ergebnis.written, "{mutation} {payload} {preview}");
            assert_eq!(ergebnis.reason_code.as_deref(), Some("invalid_binding"));
        }
        // Und es liegt nichts da — auch kein halber Beleg.
        assert!(!ort.exists() || std::fs::read_dir(&ort).unwrap().count() == 0);
    }

    #[test]
    fn ein_zweiter_beleg_ersetzt_den_ersten_statt_daneben_zu_liegen() {
        let ort = ordner("ersetzen");
        assert!(attestieren_in(&ort, MUTATION, PAYLOAD, PREVIEW, JETZT).written);
        assert!(attestieren_in(&ort, MUTATION, PAYLOAD, PREVIEW, JETZT).written);
        let anzahl = std::fs::read_dir(&ort).unwrap().count();
        assert_eq!(anzahl, 1, "es liegen {anzahl} Dateien");
    }
}
